cmd/gpu_plugin/gpu_plugin.go

```diff
@@ -283,48 +327,13 @@ func (dp *devicePlugin) scan(previousCount map[string]int) (dpapi.DeviceTree, er
 	rmDevInfos := rm.NewDeviceInfoMap()
 
 	for _, f := range files {
-		var nodes []pluginapi.DeviceSpec
-
 		if !dp.isCompatibleDevice(f.Name()) {
 			continue
 		}
 
-		drmFiles, err := os.ReadDir(path.Join(dp.sysfsDir, f.Name(), "device/drm"))
+		nodes, err := dp.getDrmFiles(f, &monitor)
 		if err != nil {
-			return nil, errors.Wrap(err, "Can't read device folder")
-		}
-
-		isPFwithVFs := pluginutils.IsSriovPFwithVFs(path.Join(dp.sysfsDir, f.Name()))
-
-		for _, drmFile := range drmFiles {
-			if dp.controlDeviceReg.MatchString(drmFile.Name()) {
-				//Skipping possible drm control node
-				continue
-			}
-
-			devPath := path.Join(dp.devfsDir, drmFile.Name())
-			if _, err := os.Stat(devPath); err != nil {
-				continue
-			}
-
-			// even querying metrics requires device to be writable
-			devSpec := pluginapi.DeviceSpec{
-				HostPath:      devPath,
-				ContainerPath: devPath,
-				Permissions:   "rw",
-			}
-
-			if !isPFwithVFs {
-				klog.V(4).Infof("Adding %s to GPU %s", devPath, f.Name())
-
-				nodes = append(nodes, devSpec)
-			}
-
-			if dp.options.enableMonitoring {
-				klog.V(4).Infof("Adding %s to GPU %s/%s", devPath, monitorType, monitorID)
-
-				monitor = append(monitor, devSpec)
-			}
+			return nil, err
```

Comment:
Can we avoid modifying scan(), please? If you need to count something in device tree please consider adding new methods to it.

Reply:
> If you need to count something in device tree please consider adding new methods to it.

Something like this: https://github.com/eero-t/intel-device-plugins-for-kubernetes/commit/5a2920854b6c26eb527c6a03b3d319e634d6f7f9  ?

It requires hard-coding the map keys upfront in Scan(), which I think is bearable.

Generalizing that would be more work: https://github.com/eero-t/intel-device-plugins-for-kubernetes/commit/4f4252ba02210b75ebfdf78cc4b0660f2ce9fcfd

(Share count is logged upfront.  Doing separate count change log messages based on `if name == "i915"` check, would IMHO be ugly.)

> Can we avoid modifying scan(), please?

As scan() is GPU plugin internal detail, why that's a problem?

(By far most of the change is scan() refactoring due to minor increase in complexity, and that would most likely be needed at some point any way.)

Reply:
> Something like this: https://github.com/eero-t/intel-device-plugins-for-kubernetes/commit/5a2920854b6c26eb527c6a03b3d319e634d6f7f9 ?

yep, that looks good to me.

Reply:
> As scan() is GPU plugin internal detail, why that's a problem?

it's not a problem, I just find https://github.com/eero-t/intel-device-plugins-for-kubernetes/commit/5a2920854b6c26eb527c6a03b3d319e634d6f7f9 much more clear.

Reply:
Ok, let's use that then.